### PR TITLE
EZP-32175: Refactor & speed up tests

### PR DESCRIPTION
**Pull request by mnocon**
JIRA: https://issues.ibexa.co/browse/EZP-32175

This PR introduces a lot of changes into how our browser tests are written while also using tactics to speed them up 
The speed up part is a bit easier to describe, so I'm going to write about it first.

# Part one: Tests speed-up

## Current state

Our current browser tests are relying mostly on actions executed by the browser. While we sometimes use API Contexts for data creation (which speeds things up) we still use the browser for navigation, so many of our tests start like this:

https://user-images.githubusercontent.com/10993858/123074561-b4590000-d417-11eb-8656-afb26272baad.mp4

It's this test: https://github.com/ezsystems/ezplatform-admin-ui/blob/master/features/standard/ContentDraft.feature#L115

It takes more than 20 seconds to navigate to the Page we want to perform actions on - we need to make sure we're on the Dashboard, click Content -> Content Structure, find the item in Subitems and only then we're done. And [it](https://github.com/ezsystems/ezplatform-admin-ui/blob/master/features/standard/ContentTypeFields.feature#L74-L76) [happens](https://github.com/ezsystems/ezplatform-form-builder/blob/master/features/FormBuilderTests.feature#L109) [a](https://github.com/ezsystems/ezplatform-page-builder/blob/master/features/DynamicLandingPage/Blocks/Blocks.feature#L52) [lot](https://github.com/ezsystems/ezplatform-admin-ui/blob/master/features/standard/Trash.feature#L67) - when navigating throught Admin interface we use the browser all the time, which seems innocent at first, but once the number of tests increases it becomes noticeable.

At first I was thinking "sure, it's slow, but it's important we make sure the navigation works!", but we don't have to check the same navigation path 1000x times during tests 😄 One Scenario is enough, the other ones can use other tricks. Which leads us to the next part...

## Things done

The solution is simple: a little less navigation, a little more action, please.

### Opening the Page that we want directly.
The new step:
```
     * @Given I'm on Content view Page for :path
     * @Given there exists Content view Page for :path
```
allows us to open a path directly, without navigating in the Browser:

https://user-images.githubusercontent.com/10993858/123080479-40b9f180-d41d-11eb-9129-62444430f608.mp4

Under the hood we parse the given path (for example: `Media/Images`) using our API, get the Content and Locations IDs and create a URL for the browser to visit (in this example: `http://example.com/admin/view/content/41/full/1/43`. This allows us to open any page in constant time (compared to the navigation method, which takes much longer for pages deeper in the tree).

There are many changes along these lines in the related PRs.

### Avoding loading the Dashboard

As you can see above we've avoided loading most of the unwanted pages on our way to the target page, but we are still visting the Dashboard. Can this be avoided?

Yes (currently it works only on OSS - I think there's an issue with following deeplinks on DXP that I need to report).
There's a new route: `/unauthenticated_login_redirect` which behaves in following way:
1. When you visit `https://example.com/admin/unauthenticated_login_redirect` as an unathenticated user
2. You are redirected to the login page
3. After successful login you're redirected to `https://example.com/admin/current_user_data`

See it in action:

https://user-images.githubusercontent.com/10993858/123093132-95646900-d42b-11eb-8399-23580543a8b7.mp4

This allows us to skip loading the Dashboard which also speeds up the tests a bit. It happens only when the shortcut step "Given I am logged as admin" is used.

### Other ideas (not implemented)
I've explored creating a session programmatically (something similar to https://github.com/symfony/symfony/pull/35997), but I was not able to get it working with Selenium, something we can explore in the future as well.

The truth is that that's almost all of the improvements I've made were like this: reduce navigation time, as most of the time there's not much else we can do. But it's certainly the direction we should aim for in the future - more API Contexts usage, less navigation, less tests that perform slow & repetitive actions.
Now the second part:

# Part two - New tests architecture

## Current state

Here are some issues with out current tests design (listed in random order)

#### 1) BrowserContext and Hooks are required in every suite.

This is a small thing, but Behat Context classes are supposed to simply provide the mapping between Step definitions and PHP code - in out stack the role of these two classes is completely different and not in line with how Behat Contexts should be used. This bad abstraction also makes it harder to write tests (what is `$this->context` and why should I pass it arround/use to to interact with the page?).

#### 2) PageObject/PageElement factories

While they allow us to achieve the goal of having most of the same test setup for different project edition they also have their downsides:
a) New Pages/Elements cannot be added in any bundle without an additional PR to Admin UI or Page Builder, which makes tests maintenence harder. It also makes it impossible to extend by others (for example: demo usage, doc usage or 3rd party usage) and created a problem when the Content edition has been created.
b) PHPStorm autocompletion does not work at all:
obraz

Correct autocompletion would display:
obraz

####  3) Our BrowserContext "API"

a) There are many methods that do similar thing (`getElementByText`, `getElementByTextFragment`, `findElement`, `getElementPositionByText`) - some of the code is duplicated, you need to know them really well to know which one can you use (and which one you shouldn't).
b) the `waitUntil` function accepts only an anonymous function, which makes it harder to reuse and wait for the same conditions in different parts of code (that's why `waitUntilElementDisappears` and `waitUntilElementIsVisible` are also present - because `waitUntil` is not reusable at all). Using this method also makes debugging tests harder, because it fails with `Spin function did not return in time` which gives you zero information on what happened.

c) `findAllElements` returns a PHP array which forces you duplicate a lot of code doing similar things (iterating over an array to find an element with given Text etc.)

d) BrowserContext became a bit of a bag for all things browser related - you mostly use it to find elements on a page, but you can also reverse match a route (`reverseMatchRoute`) or even use it in Behat Scenario as it defines a Step.

e) You can't use XPath. Our BrowserContext abstraction is based on CSS selectors and you simply cannot use anything else.

f) Operations on elements other than the root document are not easy.

you can do:
```
$baseElement = $this->context->findElement('...');
$childElement = $this->context->findElement(..., 1, $baseElement);
```

but it would be best if you could use the same API regardless of whether you're working on root document or a child element.

#### 4) Assertions

There are no built-in assertions, we use PHPUnit assertions. They work fine, but's easy to forget to write a custom failure message, which sometimes cause a failure with a message like `Failed asserting that false is true`, which is not very helpful when debugging. Example where it would happen: https://github.com/ezsystems/ezplatform-admin-ui/blob/168cf669dcd27fa9af32c433ac2b76f8c612c057/src/lib/Behat/PageElement/LeftMenu.php#L39

## New achitecture

In this PR I've created a base that all other tests can use. Below is an extensive summary on what has been changed.

#### Misc
1) I've renamed `PageObjects` to `Pages` (noone would call them `PageObjects` anyway) and `PageElements` to `Components`. `PageElements` were IMHO confusing, as by Elements we commonly refer to things that are related directly to the HTML code. `Components` became the standard name for `things that encapsulates part of the logic of the page` - for example React Components or the new Twig Components.
2) I've renamed some profiles and suites to make things consistent - we no longer have `--profile=admin-ui --suite=admin-ui`, now it's `--profile=browser --suite=admin-ui` (and similarly: `--profile=browser --suite=page-builder`). For tests not using the browser the `service` profile has been created.
3) We are slowly achieving the goal of being able to run the tests in any Symfony env, not only in `behat` - I was running them in `test` on my local environment. This will mean less maintenence for us in the future, as the `behat` env is not present by default and we need to override some recipes to make it work. The tests in other bundles are loaded based on the value of `ibexa.testing.browser.enabled` parameter - enabled by default by this bundle.
4) I've moved the Enterprise only services from this bundle to the bundles where they belong. Loading them no longer requires a special parameter - they are loaded when Workflow and Page Builder bundles are present.
5) Regression suite definitions have beem moved from page-builder to the main config files (stored here). We should slowly stop treating page-builder as the "special Behat repository" 🙃 
6) Also part of the renaming - we were strictly using the `selector` name, because our tests were only using CSS selectors. Now, as we support other means of finding elements than CSS, they are renamed to Locators, in line with the Selenium naming: https://www.browserstack.com/guide/locators-in-selenium
7) There are four main tags: `IbexaOSS`, `IbexaContent`, `IbexaExperience` and `IbexaComerce`. Each Scenario should be tagged with them to indicate which edition it's intended for.
8) The `admin` tag can still be used to use the API as admin in tests, but there's a new way: `@APIUser:<user>`, for example: `@APIUSer:admin`, which is a lot clearer.

#### Pages/Components

In order to create a new Page you can extend the [Page](https://github.com/ezsystems/BehatBundle/blob/EZP-32175-speedup-tests/src/lib/Browser/Page/Page.php) class (for Components: [Component](https://github.com/ezsystems/BehatBundle/blob/EZP-32175-speedup-tests/src/lib/Browser/Component/Component.php) class) and implement the missing methods.

For minimal examples have a look at these classes:
[Dialog](https://github.com/ezsystems/ezplatform-admin-ui/blob/EZP-32175-speedup-tests/src/lib/Behat/Component/Dialog.php)
[ChangePasswordPage](https://github.com/ezsystems/ezplatform-admin-ui/blob/EZP-32175-speedup-tests/src/lib/Behat/Page/ChangePasswordPage.php)

Let's talk about the Dialog class first. It defines two methods on its own (`accept`,`decline`) about which we will talk later (as they use our new API for interacting with the page).

The `verifyIsLoaded` method is called to make sure the HTML page is in a state that allows for further interactions - here we make sure that the dialog is loaded and visible.
The `specifyLocators` method returns an array of Locators used to find HTML elements.

There are three Locators that can be used (introduced in this PR):
1. [CSSLocator](https://github.com/ezsystems/BehatBundle/blob/EZP-32175-speedup-tests/src/lib/Browser/Locator/CSSLocator.php) - find an element using CSS
1. [XPath Locator](https://github.com/ezsystems/BehatBundle/blob/EZP-32175-speedup-tests/src/lib/Browser/Locator/XPathLocator.php#L9) - find an element using XPath
1. [VisibleCSSLocator](https://github.com/ezsystems/BehatBundle/blob/EZP-32175-speedup-tests/src/lib/Browser/Locator/VisibleCSSLocator.php#L11) - find an element using CSS **and make sure it's visible**.

Most of the time you will be using VisibleCSSLocator in the tests you write, there are only a couple of cases where you want to interact with an invisible element.

Each locator is created using the `new` keyword, in the constructor argument you pass an identifier (meant for humans, used in exception messages) and the selector.

Locators defined in the `specifyLocators` methods are available in [LocatorCollection](https://github.com/ezsystems/BehatBundle/blob/EZP-32175-speedup-tests/src/lib/Browser/Locator/LocatorCollection.php) and to refer to them you use `$this->getLocator('humanReadableLocatorIdentifier')` inside a Page/Component.

This allows us to have better error handling, for example when you make a typo in Locator name:
```
      Could not find locator with identifier: 'pageTilte'. Available identifiers are: 'pageTitle','contentType','mainContainer','tab' (InvalidArgumentException)
```

There's also a [CSSLocatorBuilder](https://github.com/ezsystems/BehatBundle/blob/EZP-32175-speedup-tests/src/lib/Browser/Locator/CSSLocatorBuilder.php) which can be useful if you want to combine several CSS locators into one.

Now let's talk about the `ChangePasswordPage`.
It introduces two new methods that come from the fact that it's a Page, not a Component:
1) `getRoute` allows us to open the Page directly (and avoid navigation, our enemy mentioned earlier in this PR)
2) `getName` allows to specify in Behat scenario the Page we want to open using a human readable name, here we would write: `Given I open "Change password page" in "admin" siteaccess` and the page would open.

As the page interacts with the right menu on the HTML page (the one with the "Publish" button) we want to use the [RightMenu](https://github.com/ezsystems/ezplatform-admin-ui/blob/EZP-32175-speedup-tests/src/lib/Behat/Component/RightMenu.php) Component. You can see that we use [constructor dependency injection](https://symfony.com/doc/current/service_container/injection_types.html#constructor-injection) for that:
```
class ChangePasswordPage extends Page
{
    /** @var \EzSystems\EzPlatformAdminUi\Behat\Component\RightMenu */
    private $rightMenu;

    public function __construct(Session $session, Router $router, RightMenu $rightMenu)
    {
        parent::__construct($session, $router);
        $this->rightMenu = $rightMenu;
    }
(...)
```
and we're free to use RightMenu methods in our class.

That's almost it 😄 We now need to register the classes as Symfony services:
https://github.com/ezsystems/ezplatform-admin-ui/blob/EZP-32175-speedup-tests/src/bundle/Resources/config/services/test/components.yaml#L16
https://github.com/ezsystems/ezplatform-admin-ui/blob/EZP-32175-speedup-tests/src/bundle/Resources/config/services/test/pages.yaml#L57

which allows us to get rid of the PageObject/PageElement factories.

And that's all - this is how you create Pages/Components and reuse existing Components in your new Pages.

Before we move to interacting with HTML pages a quick note:
since v3 we're using [FriendsOfBehat/SymfonyExtension](https://github.com/FriendsOfBehat/SymfonyExtension/) which allows us to register Behat Context classes as Symfony services.

Since Pages and Components are now services as well you can inject them into your Context classes, for example:
https://github.com/ezsystems/ezplatform-admin-ui/blob/EZP-32175-speedup-tests/src/lib/Behat/BrowserContext/UserPreferencesContext.php

#### Interacting with the HTML pages

To start interacting with the HTML page you call `$this->getHTMLPage()` inside a Page or Component.

Relevant pieces of code:
- [BaseElement](https://github.com/ezsystems/BehatBundle/blob/EZP-32175-speedup-tests/src/lib/Browser/Element/BaseElement.php), [BaseElementInterface](https://github.com/ezsystems/BehatBundle/blob/EZP-32175-speedup-tests/src/lib/Browser/Element/BaseElementInterface.php)
- [RootElement](https://github.com/ezsystems/BehatBundle/blob/EZP-32175-speedup-tests/src/lib/Browser/Element/RootElement.php), [RootElementInterface](https://github.com/ezsystems/BehatBundle/blob/EZP-32175-speedup-tests/src/lib/Browser/Element/RootElementInterface.php)

The following methods are available:
```
    public function setTimeout(int $timeoutSeconds): BaseElementInterface;
    public function find(LocatorInterface $locator): ElementInterface;
    public function findAll(LocatorInterface $locator): ElementCollection;
    public function waitUntilCondition(ConditionInterface $condition): BaseElementInterface;
    public function waitUntil(callable $callback, string $errorMessage);
    public function dragAndDrop(string $from, string $hover, string $to): void;
```

1) `setTimeout` sets the timeout value we wait until we decide that something went wrong when searching for an element on a page.
2) `find` searches the page for an element matching given locator and returns the first occurance.
3) `findAll` searches the page for all elements matching given locator and returns all occurances. They are returned as an [ElementCollection](https://github.com/ezsystems/BehatBundle/blob/EZP-32175-speedup-tests/src/lib/Browser/Element/ElementCollection.php) with which we can then interact further (ElementCollection will be discussed later with more detail).
4) `waitUntilCondition` makes sure a Condition is met before the tests execution is met. Currently there are three available conditions:
- [ElementExistsCondition](https://github.com/ezsystems/BehatBundle/blob/EZP-32175-speedup-tests/src/lib/Browser/Element/Condition/ElementExistsCondition.php) - use it when you want to make sure that an element exist on the page
- [ElementNotExistsCondition](https://github.com/ezsystems/BehatBundle/blob/EZP-32175-speedup-tests/src/lib/Browser/Element/Condition/ElementNotExistsCondition.php) - use it when you want to make sure an element has disappeared from the page (or was never present)
- [ElementCountCondition](https://github.com/ezsystems/BehatBundle/blob/EZP-32175-speedup-tests/src/lib/Browser/Element/Condition/ElementsCountCondition.php) - use to make sure the number of items matching a Locator is equal to the one you expect. Useful mostly for operations that add something to the page, for example a new Block or a Form field.
5) The `waitUntil` function is still with us 😄 In 99% of cases `waitUntilCondition` should be used as it covers most standard usages, but when you really need something custom then this function still can be used. Now it requires an error message so that we can get rid of the boring `Spin function did not return in time` exception message.
6) The `dragAndDrop` function is the improved, JS-based drag and drop that supports drag and drop operation between iframes (which Selenium does not support).

Example usages:
```
$this->getHTMLPage()->find($this->getLocator('loginButton')); // find a button
$this->getHTMLPage()->setTimeout(3)->findAll($this->getLocator('rightMenuButton')); // wait max. 3 seconds when searching for buttons matching given locator
$this->getHTMLPage()->setTimeout(10)->waitUntilCondition(new ElementExistsCondition($this->getHTMLPage(), $this->getLocator('pageTitle')); // make sure an element is present withing 10 seconds before doing other actions
$this->getHTMLPage()->setTimeout(5)->waitUntilCondition(new ElementNotExistsCondition($this->getHTMLPage(), $this->getLocator('pageBuilderBackgroundFade'))->findAll($this->getLocator('pageBuilderBlocks')); // make sure that an element is not visible before looking for elements matching locator
```

See that the last example is longer than the rest, I'd usually format it like this:
```
$this->getHTMLPage()
    ->setTimeout(5)
    ->waitUntilCondition(new ElementNotExistsCondition($this->getHTMLPage(),$this->getLocator('pageBuilderBackgroundFade'))
    ->findAll($this->getLocator('pageBuilderBlocks'));
```
for better readability.

We've found our element(s), what's next? Well, that depends on what we've found. For a single Element we have following methods ([relevant interface](https://github.com/ezsystems/BehatBundle/blob/EZP-32175-speedup-tests/src/lib/Browser/Element/ElementInterface.php) and [implementation](https://github.com/ezsystems/BehatBundle/blob/EZP-32175-speedup-tests/src/lib/Browser/Element/Element.php)):
```
    public function setTimeout(int $timeoutSeconds): BaseElementInterface;
    public function find(LocatorInterface $locator): ElementInterface;
    public function findAll(LocatorInterface $locator): ElementCollection;
    public function waitUntilCondition(ConditionInterface $condition): BaseElementInterface;
    public function waitUntil(callable $callback, string $errorMessage);
    public function isVisible(): bool;
    public function getText(): string;
    public function setValue($value): void;
    public function clear(): void;
    public function click(): void;
    public function hasAttribute(string $attribute): bool;
    public function getAttribute(string $attribute): string;
    public function hasClass(string $class): bool;
    public function mouseOver(): void;
    public function getValue();
    public function attachFile($filePath): void;
    public function getOuterHtml(): string;
    public function assert(): ElementAssert;
    public function isValid(): bool;
    public function selectOption(string $option): void;
```

The first one are the same as mentioned above. The rest of them are simple methods that interact with the page which IMHO do not need explanation - there is only one worth mentioning: `assert`.

Calling `assert()` gives you access to [ElementAssert](https://github.com/ezsystems/BehatBundle/blob/EZP-32175-speedup-tests/src/lib/Browser/Assert/ElementAssert.php) which was created to make common assertions easier.

Examples:
```
// Find an element and click it
$this->getHTMLPage()->find($this->getLocator('buttonThatIWantToClick'))->click();
// Find an element inside another element and set its value
// Timeout value is the same for both find calls
$this->getHTMLPage()
    ->setTimeout(5)
    ->find($this->getLocator('parentItem'))
    ->find($this->getLocator('childItem')
    ->setValue('Test value`);
// Find a parent element and then look inside it for all the items matching locator
$this->getHTMLPage()->find($this->getLocator('parentItem'))->findAll($this->getLocator('fieldLabel'));
// Find a parent element, make sure that it has expected Text - then search for a child element inside it, make sure it's visible and with expected text before clicking it.
$this->getHTMLPage()
    ->find($this->getLocator('myText')
    ->assert()->textEquals('I am a parent')
    ->find($this->getLocator('child'))
    ->assert()->textContains('child')
    ->assert()->isVisible()
    ->click();
```

You can see that the examples are getting more complex and complex - at this point you should be able to understand what the [Dialog](https://github.com/ezsystems/ezplatform-admin-ui/blob/EZP-32175-speedup-tests/src/lib/Behat/Component/Dialog.php) and [ChangePasswordPage](https://github.com/ezsystems/ezplatform-admin-ui/blob/EZP-32175-speedup-tests/src/lib/Behat/Page/ChangePasswordPage.php) classes (mentioned earlier) are doing.

There's only one thing left to discuss - what should you do with the result of a `findAll` call?

It returns an [ElementCollection](https://github.com/ezsystems/BehatBundle/blob/EZP-32175-speedup-tests/src/lib/Browser/Element/ElementCollection.php).

You can find following function there:
```
    public function first(): Element
    public function last(): Element
    public function count(): int
    public function any(): bool
    public function single(): Element
    public function empty(): bool

    public function map(callable $callable): array
    public function mapBy(MapperInterface $mapper): array

    public function filter(callable $callable): self
    public function filterBy(CriterionInterface $criterion): self

    public function getByCriterion(CriterionInterface $criterion): Element
    public function getBy(callable $callable): Element

    public function assert(): CollectionAssert
```

The first ones are simple:
- `first` returns the first element from all the elements found using `findAll`
- `last` returns the last element
- `count` returns how many elements were found
- `any` returns true if there's at least one element
- `single` asserts that there is **exactly** one element and returns it
- `empty` returns true when `findAll` did not find any elements

Examples:
```
// first returns a single element, so we can interact with it as described earlier
$this->getHTMLPage()->findAll($this->getLocator('buttons'))->first()->click();
// check if there are any elements matching given locator
$hasButtons = $this->getHTMLPage()->findAll($this->getLocator('button'))->any();
```

Map functions:
- the `map` function accepts an anonymous function that allows you to transform the collection into something else. Let's say you want to get the text of all elements found:
```
$elementTexts = $this->getHTMLPage()
    ->findAll($this->getLocator('button'))
    ->map(function(Element $element) { return $element->getText(); });
```

You could map the collection in any way you want, but there are some cases which are more often used than others - and getting the text of elements found is one of them.

For it you can use the [already written mapper](https://github.com/ezsystems/BehatBundle/blob/EZP-32175-speedup-tests/src/lib/Browser/Element/Mapper/ElementTextMapper.php) and the `mapBy` function. The code above can be written as:
```
$elementTexts = $this->getHTMLPage()
    ->findAll($this->getLocator('button'))
    ->mapBy(new ElementTextMapper());
```

Filter function go over all the elements in the collection and remove those that do not match given condition.
```
// Find all elements matching locator and then remove those that do not have the `button` class
$elementsThatAreButtons = $this->getHTMLPage()
    ->findAll($this->getLocator('allElements'))
    ->filter(function(Element $element) { return $element->hasClass('button'); });
```

Most of the times you will be perfoming the same filterings, so the common ones have been abstracted into [Criterions](https://github.com/ezsystems/BehatBundle/tree/EZP-32175-speedup-tests/src/lib/Browser/Element/Criterion).

Use them with the `filterBy` function:
```
$allElementsWithTextMarek = $this->getHTMLPage()
    ->findAll($this->getLocator('name'))
    ->filterBy(new ElementTextCriterion('Marek'));
```

You can filter based on the text of the element or based on the [text of it's child](https://github.com/ezsystems/BehatBundle/blob/EZP-32175-speedup-tests/src/lib/Browser/Element/Criterion/ChildElementTextCriterion.php#L13):

```
$allElementsWhoHaveChildNamedMarek = $this->getHTMLPage()
    ->findAll($this->getLocator('field'))
    ->filterBy(new ChildElementTextCriterion($this->getLocator('child'), 'Marek'));
```

This one is really common, as you can find a field based on the label that's inside it.
For example:
```
$this->getHTMLPage()
    ->findAll($this->getLocator('fields'))
    ->fiilterBy(new ChildElementTextCriterion($this->getLocator('fieldLabel'), 'My Field'))
    ->first()
    ->setValue('My Value');
```

Because calling `filterBy` returns the filtered collection you need to call `first` to get an actual element to interact with - or you can use the next group of methods: getters.

Use `get` or `getBy` the same way as you would use the corresponding `filter` function, just keep in mind that it does not return a collection - it returns the first element matching given criterion.

The same example written with `get` function:
```
$this->getHTMLPage()
    ->findAll($this->getLocator('fields'))
    ->getBy(new ChildElementTextCriterion($this->getLocator('fieldLabel'), 'My Field'))
    ->setValue('My Value');
```

And the last one, `assert`, gives you access to [CollectionAssert](https://github.com/ezsystems/BehatBundle/blob/EZP-32175-speedup-tests/src/lib/Browser/Assert/CollectionAssert.php) helping you perform collection based assertions. Currently the assertions are `isEmpty`, `hasElements`, `countEquals`.

Examples:
```
// make sure there are 5 buttons matching given Locator
$this->getHTMLPage()
    ->findAll($this->getLocator('buttons'))
    ->assert()->countEquals(5)
    ->getBy(new ElementTextCriterion('Publish')
    ->click();
```

And that's it 😄  The last ones are the most common continuations of `findAll` usages.

I know that there's a lot, so if you have any questions please let me know and I will clarify things.

### CODE REVIEW

**Comment by micszo** [comment]
> Regression suite definitions have beem moved from page-builder to the main config files (stored here). We should slowly stop treating page-builder as the "special Behat repository"

Does this mean that we will "control" test regression suites in the future via behatbundle instead of page-builder?

**Comment by mnocon** [comment]
Yes, the suite definitions are in the main config files:
- https://github.com/ezsystems/BehatBundle/blob/d1be12e2dd4dea9458b07d4552b786a8199d2f2e/behat_ibexa_oss.yaml#L70
- https://github.com/ezsystems/BehatBundle/blob/d1be12e2dd4dea9458b07d4552b786a8199d2f2e/behat_ibexa_experience.yaml#L10
- https://github.com/ezsystems/BehatBundle/blob/d1be12e2dd4dea9458b07d4552b786a8199d2f2e/behat_ibexa_content.yaml
- https://github.com/ezsystems/BehatBundle/blob/d1be12e2dd4dea9458b07d4552b786a8199d2f2e/behat_ibexa_commerce.yaml

I haven't decided yet where they will be rum, it's something I want to tackle today (it'd be best to have separate regression jobs for OSS/Content/Experience/Commerce in one repository, but that seems impossible to me right now and we will have to split it across different repos).

File: behat_ibexa_commerce.yaml [hunk 1]
```diff
@@ -2,3 +2,70 @@ imports:
...
+
+regression:
+    suites:
+        setup-commerce:
+            paths:
+                - '%paths.base%/vendor/ezsystems/behatbundle/features/setup/richtextConfiguration/custom_tags.feature'
+                - '%paths.base%/vendor/ezsystems/behatbundle/features/setup/richtextConfiguration/custom_styles.feature'
+                - '%paths.base%/vendor/ezsystems/behatbundle/features/personas'
+                - '%paths.base%/vendor/ezsystems/ezplatform-workflow/features/setup/setup_experience.feature'
+            contexts:
```
Comment (bogusez) [inline behat_ibexa_commerce.yaml:14]: can we arrange them alphabetically

File: behat_ibexa_content.yaml [hunk 2]
```diff
@@ -1,5 +1,58 @@
...
 
+regression:
+    suites:
+        setup-content:
+            paths:
+                - '%paths.base%/vendor/ezsystems/behatbundle/features/setup/richtextConfiguration/custom_tags.feature'
+                - '%paths.base%/vendor/ezsystems/behatbundle/features/setup/richtextConfiguration/custom_styles.feature'
+                - '%paths.base%/vendor/ezsystems/behatbundle/features/personas'
+                - '%paths.base%/vendor/ezsystems/ezplatform-workflow/features/setup/setup_content.feature'
+            contexts:
```
Comment (bogusez) [inline behat_ibexa_content.yaml:14]: can we arrange them alphabetically

File: behat_ibexa_experience.yaml [hunk 3]
```diff
@@ -4,3 +4,68 @@ imports:
...
+
+regression:
+    suites:
+        setup-experience:
+            paths:
+                - '%paths.base%/vendor/ezsystems/behatbundle/features/setup/richtextConfiguration/custom_tags.feature'
+                - '%paths.base%/vendor/ezsystems/behatbundle/features/setup/richtextConfiguration/custom_styles.feature'
+                - '%paths.base%/vendor/ezsystems/behatbundle/features/personas'
+                - '%paths.base%/vendor/ezsystems/ezplatform-workflow/features/setup/setup_experience.feature'
+            contexts:
```
Comment (bogusez) [inline behat_ibexa_experience.yaml:18]: can we arrange them alphabetically

File: behat_ibexa_oss.yaml [hunk 4]
```diff
@@ -57,11 +68,53 @@ default:
...
-    - vendor/ezsystems/ezplatform-user/behat_suites.yml
-
+regression:
+    suites:
+        setup-oss:
+            paths:
+                - '%paths.base%/vendor/ezsystems/behatbundle/features/setup/richtextConfiguration/custom_tags.feature'
+                - '%paths.base%/vendor/ezsystems/behatbundle/features/setup/richtextConfiguration/custom_styles.feature'
+                - '%paths.base%/vendor/ezsystems/behatbundle/features/personas'
+            contexts:
```
Comment (bogusez) [inline behat_ibexa_oss.yaml:78]: can we arrange them alphabetically

File: src/bundle/Controller/CurrentUserDataController.php [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+        $currentUserId = $this->permissionResolver->getCurrentUserReference()->getUserId();
+        $userName = $this->userService->loadUser($currentUserId)->getName();
+
+        return $this->render(
+            '@eZBehat/tests/login_data.html.twig',
+            [
+                'username' => $userName,
+                'siteaccess' => $siteaccess->name,
+            ]
```
Comment (bogusez) [inline src/bundle/Controller/CurrentUserDataController.php:44]: ];

File: src/bundle/Controller/CurrentUserDataController.php [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+        );
```
Comment (bogusez) [inline src/bundle/Controller/CurrentUserDataController.php:45]: remove ; character from this line

File: src/bundle/Resources/config/routing.yaml [hunk 7]
```diff
@@ -2,3 +2,13 @@ ez_platform_behat_repository_unauthorized_exception:
...
+
+ibexa_testing_current_user_data:
+    path: '/current_user_data'
+    defaults:
+        _controller: 'EzSystems\BehatBundle\Controller\CurrentUserDataController::showCurrentUserAction'
+
+ibexa_testing_login_redirect:
+    path: '/unauthenticated_login_redirect'
+    defaults:
+        _controller: 'EzSystems\BehatBundle\Controller\UnauthenticatedRedirectController::redirectAction'
```
Comment (bogusez) [inline src/bundle/Resources/config/routing.yaml:14]: please add an empty line at the bottom

File: src/bundle/Resources/config/services.yaml [hunk 8]
```diff
@@ -78,4 +83,10 @@ services:
     EzSystems\Behat\Core\Configuration\LocationAwareConfigurationEditor:
         decorates: EzSystems\Behat\Core\Configuration\ConfigurationEditor
 
+    EzSystems\Behat\Browser\Routing\Router:
+        arguments:
+            $router: '@ezpublish.siteaccess_router'
+
     EzSystems\Behat\Core\Configuration\ConfigurationEditorInterface: '@EzSystems\Behat\Core\Configuration\LocationAwareConfigurationEditor'
+
+    EzSystems\Behat\Browser\FileUpload\FileUploadHelper: ~
```
Comment (bogusez) [inline src/bundle/Resources/config/services.yaml:92]: please add an empty line at the bottom

File: src/bundle/Resources/views/tests/login_data.html.twig [hunk 9]
```diff
@@ -0,0 +1,6 @@
+<html>
+
+<h1> Current user is: {{ username }}</h1>
+<h1> Current siteaccess is: {{ siteaccess }}</h1>
+
+</html>
```
Comment (bogusez) [inline src/bundle/Resources/views/tests/login_data.html.twig:6]: please add an empty line at the bottom

File: dependencies.json [hunk 10]
```diff
@@ -0,0 +1,8 @@
+[
+    {
+        "requirement": "dev-EZP-32175-speedup-tests as 2.3.x-dev",
+        "repositoryUrl": "https://github.com/ezsystems/ezplatform-admin-ui",
+        "package": "ezsystems/ezplatform-admin-ui",
+        "privateRepository": false
+    }
+]
```
Comment (bogusez) [inline dependencies.json:8]: please add an empty line at the bottom

File: src/bundle/Controller/CurrentUserDataController.php (hunk 6 above)
Comment (bogusez) [inline src/bundle/Controller/CurrentUserDataController.php:45]: remove ; character here

File: behat_ibexa_commerce.yaml (hunk 1 above)
Comment (mnocon) [reply behat_ibexa_commerce.yaml:14]: Applied in https://github.com/ezsystems/BehatBundle/pull/167/commits/07de986e4f8b5c0e8462da5e16d2aca3dd2bb2a7

File: behat_ibexa_content.yaml (hunk 2 above)
Comment (mnocon) [reply behat_ibexa_content.yaml:14]: Applied in https://github.com/ezsystems/BehatBundle/pull/167/commits/07de986e4f8b5c0e8462da5e16d2aca3dd2bb2a7

File: behat_ibexa_experience.yaml (hunk 3 above)
Comment (mnocon) [reply behat_ibexa_experience.yaml:18]: Applied in https://github.com/ezsystems/BehatBundle/pull/167/commits/07de986e4f8b5c0e8462da5e16d2aca3dd2bb2a7

File: behat_ibexa_oss.yaml (hunk 4 above)
Comment (mnocon) [reply behat_ibexa_oss.yaml:78]: Applied in https://github.com/ezsystems/BehatBundle/pull/167/commits/07de986e4f8b5c0e8462da5e16d2aca3dd2bb2a7

File: dependencies.json (hunk 10 above)
Comment (mnocon) [reply dependencies.json:8]: This is an autogenerated file that will be removed before merging, there's no need to change it in any way.

File: src/bundle/Resources/config/services.yaml (hunk 8 above)
Comment (mnocon) [reply src/bundle/Resources/config/services.yaml]: Applied in https://github.com/ezsystems/BehatBundle/pull/167/commits/07de986e4f8b5c0e8462da5e16d2aca3dd2bb2a7

File: src/bundle/Resources/views/tests/login_data.html.twig (hunk 9 above)
Comment (mnocon) [reply src/bundle/Resources/views/tests/login_data.html.twig]: Applied in https://github.com/ezsystems/BehatBundle/pull/167/commits/07de986e4f8b5c0e8462da5e16d2aca3dd2bb2a7

File: src/bundle/Controller/CurrentUserDataController.php (hunk 6 above)
Comment (mnocon) [reply src/bundle/Controller/CurrentUserDataController.php:45]: I don't understand. It seems that you're suggesting to use:
```
return $this->render(
            '@eZBehat/tests/login_data.html.twig',
            [
                'username' => $userName,
                'siteaccess' => $siteaccess->name,
            ];
   )
```
instead of:
```
return $this->render(
            '@eZBehat/tests/login_data.html.twig',
            [
                'username' => $userName,
                'siteaccess' => $siteaccess->name,
            ]
   );
```

The array is passed as an argument to the render function, placing a semicolon after it would be a syntax error. Omitting the semicolon after the final closing brace would also be a syntax error.

What is it about this function that you would like to change?

Comment (bogusez) [reply src/bundle/Controller/CurrentUserDataController.php:45]: Yes, that's exactly what I meant. If that is fine then no changes are needed.

Comment (ViniTou) [reply src/bundle/Controller/CurrentUserDataController.php:45]: > If that is fine then no changes are needed.
Its not that current state is fine, is that your suggestions are syntactically invalid.

**Review by bogusez — APPROVED** [review]
